rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,427 @@
+then we would also need to bump the package of the
+[`types.LoginRuleService`](https://github.com/gravitational/teleport/blob/336518e0b51e118679c13a9a3a34dcff0fc8b5ad/api/proto/teleport/loginrule/v1/loginrule_service.proto)
+from v1 to v2.
+
+To perform the migration which allows for backward compatibility with the previous version, each step of the migration
+will be done in a phase. Each phase can be gated by a feature flag which is updated to the next phase in the process on
+subsequent Cloud releases. When all the phases have been completed the code can be cleaned up to remove the transitional
+state and feature flags required to get to the target state. The compatibility guarantees of this approach do come with
+a cost though: it will now take longer and require more process to perform the migration. We can however mitigate the
+time required to progress through each phase by increasing the Cloud release cadence.
+
+| Phase | Write                  | Read                                        | Details                                                                       |
+| ----- | ---------------------- | ------------------------------------------- | ----------------------------------------------------------------------------- |
+| 1     | Old key AND new key(s) | Old Key                                     | Introduce new resource(s) and mirror data to new key range on write           |
+| 2     | Old key AND new key(s) | New Key, fallback to old key if not present | Start reading from the new key and using the old key as a fall back           |
+| 3     | Old key AND new key(s) | New Key, fallback to old key if not present | Run data migration that copies all keys from old to new                       |
+| 4     | New key(s)             | New Key                                     | Stop mirroring writes and falling back to old resources when they don't exist |
+| 5     | New key(s)             | New Key                                     | Run optional migration that removes all old keys if they do not have a TTL    |
+
+#### Data Migration Phase
+
+Data migrations no longer need to occur during Auth initialization, which will reduce the time it will take a new Auth
+instance to start processing requests. By waiting until Phase 3 to run data migrations when resource mirroring and
+fallback reads are already being performed by all Auth instances the migration can happen in the background.
+
+This phase is only required for resources which do not heartbeat. For any presence related resources we can use the
+write caused by the next heartbeat to perform the data migration.
+
+> The above only applies when doing phased migrations. Major releases used by self-hosted customers will still be using
+> the one-shot migration approach will still require running the migration during initialization.
```

Comment:
This feels like a point worth expounding upon.  IIUC this means that any implementation requiring a migration will need to be written s.t. teleport can intelligently decide wether the migration should be blocking or asynchronous.  IMO this RFD should provide a specific recommendation/framework for how that should be handled.

Reply:
Ah, I hadn't understood that we were literally making this behavior cloud-specific.  Got it.